Source/websocket/JSONRPCLink.h

```diff
@@ -1261,30 +1330,77 @@ namespace Thunder {
                 ~Connection() override
                 {
                     _monitor.Revoke(_T("statechange"));
+                    _eventSubscriber.Stop();
+                    _eventSubscriber.Wait(Core::Thread::STOPPED, Core::infinite);
                 }
 
             public:
                 bool IsActivated()
                 {
                     return (_state == ACTIVATED);
                 }
+                void SubscribeEvents() {
+                    _adminLock.Lock();
+                    for (const string& iter: _subscriptions) {
+                        SendSubscribeRequest(iter);
+                    }
+                    _adminLock.Unlock();
+                    _eventSubscriber.Block();
+                    _state = state::ACTIVATED;
+                    _parent.StateChange();
+                }
+
+            template <typename INBOUND, typename METHOD>
+            uint32_t Subscribe(const uint32_t waitTime, const string& eventName, const METHOD& method)
+            {
+                auto result = Base::template Subscribe<INBOUND, METHOD>(waitTime, eventName, method);
+                if (result == Core::ERROR_NONE) {
+                    _adminLock.Lock();
+                    _subscriptions.insert(string(eventName));
+                    _adminLock.Unlock();
+                }
+                return result;
+            }
+            template <typename INBOUND, typename METHOD, typename REALOBJECT>
+            uint32_t Subscribe(const uint32_t waitTime, const string& eventName, const METHOD& method, REALOBJECT* objectPtr)
+            {
+                auto result = Base::template Subscribe<INBOUND, METHOD, REALOBJECT>(waitTime, eventName, method, objectPtr);
+                if (result == Core::ERROR_NONE) {
+                    _adminLock.Lock();
+                    _subscriptions.insert(string(eventName));
+                    _adminLock.Unlock();
+                }
+                return result;
+            }
+            void Unsubscribe(const uint32_t waitTime, const string& eventName)
+            {
+                _adminLock.Lock();
+                ASSERT(_subscriptions.find(eventName) != _subscriptions.end());
+                auto iter = _subscriptions.erase(eventName);
+                _adminLock.Unlock();
+                return Base::Unsubscribe(waitTime, eventName);
+            }
+
 
             private:
+                uint32_t SendSubscribeRequest(const string& eventName) {
+                    uint32_t retVal = Core::ERROR_UNAVAILABLE;
+                    Core::JSONRPC::Message response;
+                    const string parameters("{ \"event\": \"" + eventName + "\", \"id\": \"" + Base::Namespace() + "\"}");
+                    auto result = Base::template Invoke<string>(DefaultWaitTime, "register", parameters, response);
```

Comment:
Oke, synchronous Invoke, so communication thread is not an option :-) Maybe than 1 thread for all processes that have this ?